ODIN_II/SRC/netlist_create_from_ast.cpp

```diff
@@ -3469,51 +3472,213 @@ signal_list_t* create_operation_node(ast_node_t* op, signal_list_t** input_lists
+                    for (i = 0; i < output_port_width - shift_size; i++) {
+                        /* hook the output pin into the node */
+                        add_output_pin_to_node(operation_node, operation_node->input_pins[i], i + shift_size);
+                        // connect higher output pin to lower input pin
+                        add_pin_to_signal_list(return_list, operation_node->input_pins[i]);
+                        // set NULL as node of input_pins
+                        operation_node->input_pins[i]->node = NULL;
+                    }
+                    break;
+                }
+                case ASL: 
+                {
+                    /* record the size of the shift */
+                    int shift_size = op->children[1]->types.vnumber->get_value();               
+
+                    /* connect ZERO to outputs that don't have inputs connected */
+                    for (i = 0; i < shift_size; i++) {
+                        // connect 0 to lower outputs
+                        npin_t* zero_pin = allocate_npin();
+                        add_fanout_pin_to_net(verilog_netlist->zero_net, zero_pin);
+                        /* hook the output pin into the node */
+                        add_output_pin_to_node(operation_node, zero_pin, i);
+
+                        add_pin_to_signal_list(return_list, zero_pin);
+                        zero_pin->node = NULL;
+                    }
+
+                    /* connect inputs to outputs */
+                    for (i = 0; i < output_port_width - shift_size; i++) {
+                        /* hook the output pin into the node */
+                        add_output_pin_to_node(operation_node, operation_node->input_pins[i], i + shift_size);
+                        // connect higher output pin to lower input pin
+                        add_pin_to_signal_list(return_list, operation_node->input_pins[i]);
+                        // set NULL as node of input_pins
+                        operation_node->input_pins[i]->node = NULL;
+                    }
+                    break;
+                }
+                case SR: 
+                {
+                    /* record the size of the shift */
+                    int shift_size = op->children[1]->types.vnumber->get_value(); 
+
+                    for (i = shift_size; i < output_port_width; i++) {
+                        /* hook the output pin into the node */
+                        add_output_pin_to_node(operation_node, operation_node->input_pins[i], i - shift_size);
+                        // connect higher output pin to lower input pin
+                        add_pin_to_signal_list(return_list, operation_node->input_pins[i]);
+                        // set NULL as node of input_pins
+                        operation_node->input_pins[i]->node = NULL;
+                    }
+
+                    /* connect ZERO to outputs that don't have inputs connected */
```

Comment:
merge ASR and SR logic, they are the same except the pad

Reply:
What I understood from the IEEE Standard is that we need to know is the input signed or not. There is not only extend the MSB. However, it may be a false conclusion. Here is the link, please notify me again about your final decision.

http://staff.ustc.edu.cn/~songch/download/IEEE.1364-2005.pdf#G6.9007

Reply:
you are right it seems!
I believe there is a flag for is_signed. so use that and put an oassert(false && "TODO: test that ASR works with signed number") for example